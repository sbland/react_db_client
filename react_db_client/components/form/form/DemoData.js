import { filterTypes } from '@react_db_client/constants.client-types';

export const demoHeadingsData = [
  {
    uid: 'uid',
    label: 'UID',
    type: filterTypes.uid,
  },
  {
    uid: 'name',
    label: 'Text',
    type: filterTypes.text,
    required: true,
  },
  {
    uid: 'longText',
    label: 'Long Text',
    type: filterTypes.textLong,
  },
  {
    uid: 'date',
    label: 'Date',
    type: filterTypes.date,
  },
  {
    uid: 'number',
    label: 'Number',
    type: filterTypes.number,
  },
  {
    uid: 'number-capped',
    label: 'Number Capped',
    type: filterTypes.number,
    min: 3,
    max: 8,
    step: 0.1,
    defaultValue: 5,
  },
  {
    uid: 'demoField',
    label: 'Custom Field Types',
    type: 'demoFieldType',
  },
  {
    uid: 'reference',
    label: 'Reference',
    collection: 'democollection',
    type: filterTypes.reference,
    objectLink: true,
  },
  {
    uid: 'file',
    label: 'File',
    collectionId: 'democollection',
    documentId: 'docid',
    fileType: 'document',
    type: filterTypes.file,
  },
  {
    uid: 'image',
    label: 'Image',
    type: filterTypes.image,
  },
  {
    uid: 'bool',
    label: 'Bool',
    type: filterTypes.bool,
  },
  {
    uid: 'select',
    label: 'Select',
    type: filterTypes.select,
    options: [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectreadonly',
    label: 'Select Read Only',
    type: filterTypes.select,
    readOnly: true,
    options: [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectSearch',
    label: 'Select Search',
    type: filterTypes.selectSearch,
    searchFieldTargetField: 'label',
    searchFn: async () => [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectSearchMulti',
    label: 'Select Search Multi',
    type: filterTypes.selectSearch,
    multiple: true,
    searchFieldTargetField: 'label',
    allowEmptySearch: true,
    searchFn: async () => [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'multiSelect',
    label: 'Multi Select',
    type: filterTypes.selectMulti,
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'multiSelectlist',
    label: 'Multi Select No Dropdown',
    type: filterTypes.selectMulti,
    asDropdown: false,
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'multiSelectlistShowAll',
    label: 'Multi Select Show All',
    type: filterTypes.selectMulti,
    selectType: 'showall',
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'embedded',
    label: 'Embedded',
    type: filterTypes.embedded,
    children: [{ uid: 'embeddedtext', label: 'Embedded Text', type: filterTypes.text }],
  },
  {
    uid: 'embeddedb',
    label: 'Embedded B',
    type: filterTypes.embedded,
    orientation: 'horiz',
    children: [
      { uid: 'embeddedtog1', label: 'Embedded Tog1', type: filterTypes.bool },
      { uid: 'embeddedtog2', label: 'Embedded Tog2', type: filterTypes.bool },
      { uid: 'embeddedtog3', label: 'Embedded Tog3', type: filterTypes.bool },
    ],
  },
  {
    uid: 'dict',
    label: 'Dictionary Field',
    type: filterTypes.dict,
  },
  {
    uid: 'video',
    label: 'Video     Field',
    type: filterTypes.video,
  }
];

export const demoFormData = {
  name: 'Name 1',
  uid: 'name-1',
  date: '2019-11-02T12:04:44.626+00:00',
  selectreadonly: 'rep1',
  multiSelect: ['foo', 'bar'],
  longText: `Long Text spanning multiple lines
  1
  2
  3
  4
  5
  6
  7
  8
  9
  `,
};

export const demoAdditionalData = {
  customFieldStyle: { background: 'grey', padding: '2px' },
};
