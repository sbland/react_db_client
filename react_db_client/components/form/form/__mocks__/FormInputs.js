import React from 'react';
import PropTypes from 'prop-types';


const FormInputs = ({
  headings,
  formData,
  updateFormData,
  orientation,
  heading,
  isSection,
}) => {
  const className = `
    form_inputs
    ${(isSection) ? 'formSection' : ''}
    ${orientation}
  `;
  return (
    <section className={className}>
      {(!isSection) && <p>* is required.  (!) has been Overriden.</p>}
      <h4 className="formSection_heading">{heading}</h4>
      <div className="form_row">
        <input type="text" />
      </div>
    </section>
  );
};

FormInputs.propTypes = {
  headings: PropTypes.arrayOf(PropTypes.shape({
    uid: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    type: PropTypes.oneOf([
      'uid',
      'reference',
      'file',
      'fileMultiple',
      'textLong',
      'text',
      'number',
      'date',
      'image',
      'bool',
      'select',
      'selectMulti',
      'embedded',
    ]).isRequired,
    unit: PropTypes.string,
  })).isRequired,
  formData: PropTypes.object.isRequired,
  updateFormData: PropTypes.func.isRequired,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  heading: PropTypes.string,
  isSection: PropTypes.bool,
};

FormInputs.defaultProps = {
  orientation: 'vert',
  heading: '',
  isSection: false,
};

export default FormInputs;
