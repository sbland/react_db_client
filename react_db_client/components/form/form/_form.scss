@import '@samnbuk/react_db_client.constants.style';
$section-padding: $unit * 0.1;
$section-border: 1px $white dashed;
$heading-height: $unit;

.form,
.form_inputs,
.form_section {
  /* FORM SECTION */
  &.formSection,
  .formSection {
    position: relative;
    display: flex;
    overflow: visible;
    flex-grow: 1;

    // Styling
    padding: $section-padding;
    padding-top: $section-padding; // Allows for title
    padding-bottom: $section-padding;
    border: $section-border;
    border-radius: ($unit * 0.3);

    margin: 0; // Reset margin
    margin-top: $section-padding;
    box-shadow: 0px 0px 10px 0 rgba(0, 0, 0, 0.2);

    &.hasHeading {
      padding-top: $section-padding + $heading-height; // Allows for title
      margin-top: $section-padding + $heading-height;
    }
  }
  // Child of vert section
  &.vert > .formSection {
    // border: 3px red solid;
    // margin-top: $section-padding * 2;
    // padding-top: $section-padding;
    &:not(:last-child) {
      margin-bottom: $section-padding;
    }
  }
  // Child of horiz section
  &.horiz > .formSection {
    // border: 3px green solid;
    margin-left: $section-padding;
    &:last-child {
      margin-right: $section-padding;
    }
  }
  // vert section
  &.formSection.vert,
  .formSection.vert {
    // background: blue;
    flex-direction: column;
    // padding-top: $section-padding;
  }
  // horiz section
  &.formSection.horiz,
  .formSection.horiz {
    // background: yellow;
    padding-top: $section-padding;
    align-items: stretch;
    justify-items: center;
    justify-content: space-between;
    flex-wrap: wrap;
  }

  /* FORM ROW */
  .form_row {
    flex-grow: 1;
    display: flex;
    margin: 0; // Reset margin
  }
  &.vert > .form_row {
    // background: grey;
    align-items: flex-end; // Ensures inputs align to bottom of label
    &:not(:first-child) {
      margin-top: $section-padding;
    }

    &:not(:last-child) {
      margin-bottom: $section-padding;
    }
  }
  &.horiz > .form_row {
    // background: fuchsia;
    flex-direction: column;
    margin-right: $section-padding;
    justify-content: flex-start;
    &:not(:first-child) {
      margin-left: $section-padding;
    }
  }

  /* FORM LABEL */
  .form_label {
    text-align: right;
    // flex-grow: 1;
    margin-right: $unit;
    &.required {
      color: $warningColour;
    }

    &.hasChanged {
      // color: $warningColour;
      position: relative;
      &::after {
        content: '';
        position: absolute;
        display: block;
        width: 2px;
        background: $warningColour;
        height: 100%;
        z-index: 10;
        right: -$unit;
        top: 0;
      }
    }
  }
  &.vert > .form_row > .form_label {
    width: 30%;
    max-width: 200px;
    flex-grow: 1;
  }
  &.horiz > .form_row > .form_label {
    text-align: right;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: fit-content;
    width: fit-content;
  }

  /* SECTION HEADING */

  .formSection_heading {
    position: absolute;
    background: whitesmoke;
    top: -$unit * 0.5;
    left: $unit;
    width: auto;
    border-radius: 3px;
  }
  h1,
  h2,
  h3,
  h4 {
    transition: all 0.2s ease-out;
  }

  &:hover {
    h1,
    h2,
    h3,
    h4 {
      color: $primaryColour;
    }
    & > .formSection_heading {
      background: white;
    }
    // .formSection {
    //   .formSection_heading {
    //     background: whitesmoke;
    //   }
    // }
  }

  /* INPUTS */

  input[type='text'],
  input[type='number'] {
    min-width: 60%;
    // height: 100%;
    flex-grow: 1;
    // max-height: 2.5rem;
    border: 1px $grey-30 solid;
    // outline: none;

    &:focus {
      // border: none;
      // border: 1px $primaryColour solid;
      // outline: 1px $primaryColour solid;
    }
  }

  select {
    min-width: 60%;
    // height: 100%;
    // max-height: 2.5rem;
    border: 1px $grey-30 solid;
    outline: none;
  }

  .formFieldFile {
    border: 1px $primaryColour solid;
    background: $grey-20;
  }

  .formFieldInput {
    .vert & {
      flex-grow: 1;
    }
    .horiz & {
      flex-grow: 0;
    }
  }

  textarea {
    width: 100%;
    flex-grow: 1;
  }

  input[type='checkbox'],
  input[type='radio'] {
    margin: 0 ($unit * 0.5);
  }

  .submitBtns {
    display: flex;
    width: 100%;
  }

  .formFieldInput {
    width: 100%;
  }
}
