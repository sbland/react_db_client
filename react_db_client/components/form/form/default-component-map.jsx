import React from 'react';
import { filterTypes } from '@samnbuk/react_db_client.constants.client-types';

import { FieldText } from '@samnbuk/react_db_client.components.form.form-fields.field-text';
import { FieldTextArea } from '@samnbuk/react_db_client.components.form.form-fields.field-text-area';
import { FieldBool } from '@samnbuk/react_db_client.components.form.form-fields.field-bool';
import { FieldDate } from '@samnbuk/react_db_client.components.form.form-fields.field-date';
import { FieldNumber } from '@samnbuk/react_db_client.components.form.form-fields.field-number';
import { FieldObjectRef } from '@samnbuk/react_db_client.components.form.form-fields.field-object-ref';
import { FieldMultiSelect } from '@samnbuk/react_db_client.components.form.form-fields.field-multi-select';
import { FieldSelect } from '@samnbuk/react_db_client.components.form.form-fields.field-select';
import { FieldSelectSearch } from '@samnbuk/react_db_client.components.form.form-fields.field-select-search';
import { FieldFile } from '@samnbuk/react_db_client.components.form.form-fields.field-file';

export const defaultComponentMap = ({ asyncGetDocuments, fileServerUrl } = {}) => ({
  [filterTypes.text]: () => FieldText,
  [filterTypes.select]: () => FieldSelect,
  [filterTypes.selectMulti]: () => FieldMultiSelect,
  // [filterTypes.fileMultiple]: () => FieldFile,
  [filterTypes.file]: () => (props) =>
    <FieldFile {...props} fileServerUrl={fileServerUrl} asyncGetDocuments={asyncGetDocuments} />,
  // [filterTypes.image]: () => FieldFile,
  [filterTypes.textLong]: () => FieldTextArea,
  [filterTypes.number]: () => FieldNumber,
  [filterTypes.date]: () => FieldDate,
  [filterTypes.bool]: () => FieldBool,
  [filterTypes.toggle]: () => FieldBool,
  [filterTypes.selectSearch]: () => FieldSelectSearch,
  [filterTypes.reference]: () => (props) =>
    <FieldObjectRef {...props} asyncGetDocuments={asyncGetDocuments} />,
});
