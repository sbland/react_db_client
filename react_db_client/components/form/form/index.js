export { Form } from './form';
export { FormInputs } from './FormInputs';
export * as FormStyles from './_form.scss';
export { defaultComponentMap } from './default-component-map';
export { FormField } from './FormField';