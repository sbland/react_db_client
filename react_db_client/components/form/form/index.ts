export { Form } from './form';
export { FormInputs } from './FormInputs';
