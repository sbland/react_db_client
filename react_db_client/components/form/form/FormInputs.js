import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { filterTypes } from '@react_db_client/constants.client-types';

import '@samnbuk/react_db_client.constants.style';
import { FormField as DefaultFormField } from './FormField';
import './_form.scss';

export const FormInputs = ({
  FormField,
  headings,
  formData,
  updateFormData,
  orientation,
  heading: sectionHeading,
  isSection,
  showKey,
  additionalData,
  componentMap,
  id,
}) => {
  const className = [
    'form_inputs',
    'formSection',
    `${orientation}`,
    sectionHeading ? 'hasHeading' : '',
    `formSection_${id}`,
  ]
    .filter((f) => f)
    .join(' ');

  const fields = useMemo(
    () =>
      headings.map((heading) => {
        const value =
          formData && formData[heading.uid] !== null
            ? formData[heading.uid]
            : heading.defaultValue || heading.def;

        if (heading.type === filterTypes.embedded) {
          return (
            <FormInputs
              key={heading.uid}
              headings={heading.children}
              formData={formData}
              updateFormData={updateFormData}
              orientation={heading.orientation}
              heading={heading.label}
              additionalData={additionalData}
              componentMap={componentMap}
              FormField={FormField}
              id={heading.uid}
              isSection
            />
          );
        }
        // return <div>{heading.uid}</div>;
        return (
          <FormField
            heading={heading}
            updateFormData={updateFormData}
            value={value}
            key={heading.uid}
            additionalData={additionalData}
            componentMap={componentMap}
          />
        );
      }),
    [headings, formData, updateFormData, additionalData, componentMap]
  );

  return (
    <section className={className}>
      {!isSection && showKey && <p>* is required. (!) has been modified.</p>}
      <h4 className="formSection_heading">{sectionHeading}</h4>
      {fields}
    </section>
  );
};

FormInputs.propTypes = {
  FormField: PropTypes.func,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      orientation: PropTypes.oneOf(['horiz', 'vert']),
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      unit: PropTypes.string,
      children: PropTypes.arrayOf(PropTypes.shape({})),
    })
  ).isRequired,
  formData: PropTypes.object.isRequired,
  updateFormData: PropTypes.func.isRequired,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  heading: PropTypes.string,
  isSection: PropTypes.bool,
  showKey: PropTypes.bool,
  additionalData: PropTypes.shape({}),
  componentMap: PropTypes.objectOf(PropTypes.elementType).isRequired,
  id: PropTypes.any,
};

FormInputs.defaultProps = {
  FormField: DefaultFormField,
  orientation: 'vert',
  heading: '',
  isSection: false,
  showKey: true,
  additionalData: {},
  id: null,
};
