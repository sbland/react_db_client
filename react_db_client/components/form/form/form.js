import React, { useCallback, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

import cloneDeep from 'lodash/cloneDeep';

import { FormInputs } from './FormInputs';

import './_form.scss';
import formValidation from './formValidation';
// import Emoji from '../Emoji/Emoji';

/** Generic Form Component that manages updates and layout
 *
 * The Form component handles general form requirements such as storing
 * edit data, auto generating a layout bassed on provided headers.


 # Headings
 The headings should have the correct headings shape...

 # Mounting the save buttons using a ref:
 If a react ref is provided to `endButtonRefOverride` then the save buttons will be placed in
 the referenced container.


 * On submit signature:
 * ({ formEditData, formData }) => {...}
 *
 * onChange Signature:
 * (field, value, newFormData) => {...}


 * additionalData - additional data to be passed to form field components. Useful when using a custom field
 * customFieldComponents - a map of field type against react component
 */
export const Form = ({
  FormField,
  formDataInitial,
  headings,
  onSubmit,
  onChange,
  showEndBtns,
  submitBtnText,
  showKey,
  orientation,
  disableAutocomplete,
  endButtonRefOverride,
  errorCallback,
  additionalData,
  customFieldComponents,
}) => {
  // const [formData, setFormData] = useState(cloneDeep(formDataInitial));
  const [formEditData, setFormEditData] = useState({});
  const [endButtonContainerRef, setEndButtonContainerRef] = useState(null);

  const formData = useMemo(() => ({ ...cloneDeep(formDataInitial), ...formEditData }), [
    formDataInitial,
    formEditData,
  ]);

  const updateFormData = useCallback(
    (field, value) => {
      setFormEditData((prev) => {
        const newFormData = cloneDeep(prev);
        newFormData[field] = value;
        if (onChange) onChange(field, value, newFormData);
        return newFormData;
      });
    },
    [onChange]
  );

  const handleReset = () => {
    setFormEditData({});
  };

  const handleSubmit = useCallback(() => {
    const passesFormValidation = formValidation(formData, headings);
    if (passesFormValidation === true) onSubmit({ formEditData, formData });
    else errorCallback(passesFormValidation.error);
  }, [formData, formEditData, headings, onSubmit, errorCallback]);

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        handleSubmit();
      }}
      // Another possible hack to stop autocomplete on chrome
      // autoComplete={disableAutocomplete ? 'chrome-off' : 'on'}
      autoComplete={disableAutocomplete ? 'off' : 'on'}
      className="form sectionWrapper"
    >
      <FormInputs
        headings={headings}
        formData={formData}
        updateFormData={updateFormData}
        showKey={showKey}
        orientation={orientation}
        additionalData={additionalData}
        customFieldComponents={customFieldComponents}
        FormField={FormField}
      />
      <section ref={(ref) => setEndButtonContainerRef(ref)} style={{ width: '100%' }} />
      {showEndBtns &&
        endButtonContainerRef &&
        ReactDOM.createPortal(
          <div className="submitBtns">
            {/* TODO: Work out why setting this as a submit button does not work */}
            <button type="button" className="button-two submitBtn" onClick={handleSubmit}>
              <span role="img" aria-label={submitBtnText}>
                üíæ
              </span>
            </button>
            <button type="button" className="button-one" onClick={handleReset}>
              <span role="img" aria-label="Delete">
                ‚ùå
              </span>
            </button>
          </div>,
          endButtonRefOverride || endButtonContainerRef
        )}
    </form>
  );
};

Form.propTypes = {
  formDataInitial: PropTypes.shape().isRequired,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.oneOf([
        'uid',
        'reference',
        'file',
        'textLong',
        'text',
        'number',
        'date',
        'image',
        'bool',
        'select',
        'selectMulti',
        'embedded',
      ]).isRequired,
    })
  ).isRequired,
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func,
  showEndBtns: PropTypes.bool,
  submitBtnText: PropTypes.string,
  disableAutocomplete: PropTypes.bool,
  showKey: PropTypes.bool,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  endButtonRefOverride: PropTypes.any, // Should be a react reference
  errorCallback: PropTypes.func,
  additionalData: PropTypes.shape({}),
  customFieldComponents: PropTypes.objectOf(PropTypes.elementType),
};

Form.defaultProps = {
  onChange: () => {},
  showEndBtns: true,
  submitBtnText: 'Save',
  showKey: true,
  orientation: 'vert',
  disableAutocomplete: false,
  endButtonRefOverride: null,
  errorCallback: alert,
  additionalData: {},
  customFieldComponents: {},
};

