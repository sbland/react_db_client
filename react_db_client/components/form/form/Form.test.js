import '@samnbuk/react_db_client.helpers.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Form } from './form';
import { demoHeadingsData, demoFormData } from './DemoData';
import { FormInputs } from './FormInputs';
import { defaultComponentMap } from './default-component-map';

const submitFunc = jest.fn();
const onChangeFunc = jest.fn();

// gets the number of rows
const rowCounter = (acc, val) => {
  let addme = 0;
  if (val.children) {
    addme = val.children.reduce(rowCounter, 0);
  } else {
    addme += 1;
  }
  return acc + addme;
};

const FormField = () => <div>FormFieldMock</div>;

const defaultProps = {
  headings: demoHeadingsData,
  formDataInitial: demoFormData,
  onSubmit: submitFunc,
  componentMap: defaultComponentMap,
  FormField,
};

describe('Form', () => {
  it('renders', () => {
    shallow(<Form {...defaultProps} />);
  });

  it('Matches Snapshot', () => {
    const component = shallow(<Form {...defaultProps} />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });

  const form = mount(<Form {...defaultProps} />);

  it('renders a row for each heading', () => {
    expect(form.find(FormField).length).toEqual(demoHeadingsData.reduce(rowCounter, 0));
  });

  it('renders embedded headings as sections', () => {
    // expect(form.find())
  });

  it.skip('renders a submit button', () => {
    // Skipping as the submit button is now moved to a ref
    // TODO: May fail if more buttons on form
    expect(form.find('.submitBtn').props().type).toEqual('submit');
  });

  describe('submit', () => {
    it('calls on submit on submit button press', () => {
      const formDOM = form.find('.form');
      formDOM.simulate('submit');
      expect(submitFunc).toHaveBeenCalledTimes(1);
      expect(submitFunc).toHaveBeenCalledWith({
        formData: demoFormData,
        formEditData: {},
      });
    });

    it('Submits with new data when input changed and submit button pressed', () => {
      const formDOM = form.find('.form');
      // Disabled ids on form inputs to prevent autofill
      // const formInputDOM = form.find('input').filterWhere((n) => n.props().id === 'name');
      const formInputDOM = form.find('input').first();
      formInputDOM.simulate('change', { target: { value: 'foo' } });
      formDOM.simulate('submit');
      expect(submitFunc).toHaveBeenCalledTimes(2);
      expect(submitFunc).lastCalledWith({
        formData: { ...demoFormData, name: 'foo' },
        formEditData: { name: 'foo' },
      });
    });
  });

  describe('handleChange', () => {
    it('calls onChange func when a field is changed during live update', () => {
      const formLive = mount(<Form {...defaultProps} onChange={onChangeFunc} />);
      const formInputDOM = formLive
        .find('input')
        // Disabled ids on form inputs to prevent autofill
        // .filterWhere((n) => n.props().id === 'name')
        .first();
      formInputDOM.simulate('change', { target: { value: 'foo' } });
      expect(onChangeFunc).toHaveBeenCalledWith('name', 'foo', { name: 'foo' });
    });
  });
});

describe('FormInputs', () => {
  test.only('should update on change', () => {
    const mockUpdate = jest.fn();
    const formInputs = mount(
      <FormInputs {...defaultProps} updateFormData={mockUpdate} formData={{}} />
    );
    const formField = formInputs.find(FormField).first();

    act(() => {
      formField.props().updateFormData('name', 'foo');
    });
    expect(mockUpdate).toHaveBeenCalledWith('name', 'foo');
  });
});
