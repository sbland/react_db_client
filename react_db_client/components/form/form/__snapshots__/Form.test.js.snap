// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Matches Snapshot 1`] = `
"<form onSubmit={[Function: onSubmit]} autoComplete=\\"on\\" className=\\"form sectionWrapper\\">
  <FormInputs headings={{...}} formData={{...}} updateFormData={[Function]} showKey={true} orientation=\\"vert\\" additionalData={{...}} customFieldComponents={{...}} heading=\\"\\" isSection={false} />
  <section style={{...}} />
</form>"
`;
